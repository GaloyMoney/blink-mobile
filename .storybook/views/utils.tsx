import { useThemeMode } from "@rneui/themed"
import React from "react"

export const SetDarkMode: React.FC = () => {
    const { mode, setMode } = useThemeMode()
    React.useEffect(() => {
      console.log("mode", mode)
      if (mode === "dark") return
      setMode("dark")
    }, [setMode, mode])
    return null
  }

export const SetLightMode: React.FC = () => {
    const { mode, setMode } = useThemeMode()
    React.useEffect(() => {
      console.log("mode", mode)
      if (mode === "light") return
      setMode("light")
    }, [setMode, mode])
    return null
  }