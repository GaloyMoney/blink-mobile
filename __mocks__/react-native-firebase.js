/* eslint-disable no-undef */
jest.mock("react-native-firebase", () => {
  return {
    default: {
      messaging: jest.fn(() => ({
        hasPermission: jest.fn(() => Promise.resolve(true)),
        subscribeToTopic: jest.fn(),
        unsubscribeFromTopic: jest.fn(),
        requestPermission: jest.fn(() => Promise.resolve(true)),
        getToken: jest.fn(() => Promise.resolve("myMockToken")),
      })),
      notifications: jest.fn(() => ({
        onNotification: jest.fn(),
        onNotificationDisplayed: jest.fn(),
      })),
      analytics: jest.fn(() => ({
        logEvent: jest.fn(),
      })),
      functions: jest.fn(() => ({
        httpsCallable: jest.fn(),
      })),
    },
  }
})
