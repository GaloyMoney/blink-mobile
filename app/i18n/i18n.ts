import i18n from "i18n-js"

import * as en from "./en.json"
import * as es from "./es.json"

i18n.fallbacks = true
i18n.translations = { en, es }
