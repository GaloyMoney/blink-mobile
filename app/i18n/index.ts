// eslint-disable-next-line import/no-unassigned-import
import "./i18n"

export * from "./translate"
