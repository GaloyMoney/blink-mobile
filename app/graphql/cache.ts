import { InMemoryCache, gql } from "@apollo/client"
import { Account, MyWalletsFragmentDoc, Wallet, WalletCurrency } from "./generated"
import { relayStylePagination } from "@apollo/client/utilities"
import { ReadFieldFunction } from "@apollo/client/cache/core/types/common"

gql`
  fragment MyWallets on ConsumerAccount {
    wallets {
      id
      balance
      walletCurrency
    }
  }

  query realtimePrice {
    me {
      id
      defaultAccount {
        id
        realtimePrice {
          btcSatPrice {
            base
            offset
          }
          denominatorCurrency
          id
          timestamp
          usdCentPrice {
            base
            offset
          }
        }
      }
    }
  }
`

type getWalletsInputs = {
  readField: ReadFieldFunction
  cache: InMemoryCache
}

const getWallets = ({
  readField,
  cache,
}: getWalletsInputs): readonly Wallet[] | undefined => {
  const id = readField("id")
  const key = `ConsumerAccount:${id}`
  const account: Account | null = cache.readFragment({
    id: key,
    fragment: MyWalletsFragmentDoc,
  })
  if (account === null) {
    return undefined
  }
  return account.wallets
}

export const createCache = () =>
  new InMemoryCache({
    possibleTypes: {
      // TODO: add other possible types
      Wallet: ["BTCWallet", "UsdWallet"],
      Account: ["ConsumerAccount"],
    },
    typePolicies: {
      Globals: {
        // singleton: only cache latest version:
        // https://www.apollographql.com/docs/react/caching/cache-configuration/#customizing-cache-ids
        keyFields: [],
      },
      RealtimePrice: {
        keyFields: [],
      },
      MapMarker: {
        keyFields: ["mapInfo", ["title", "coordinates"]],
      },
      Contact: {
        fields: {
          prettyName: {
            read(_, { readField }) {
              return readField("id") || readField("name")
            },
          },
        },
      },
      UserContact: {
        fields: {
          transactions: relayStylePagination(),
        },
      },
      Earn: {
        fields: {
          completed: {
            read: (value) => value ?? false,
          },
        },
      },
      Query: {
        fields: {
          // local only fields
          hideBalance: {
            read: (value) => value ?? false,
          },
          hiddenBalanceToolTip: {
            read: (value) => value ?? false,
          },
          beta: {
            read: (value) => value ?? false,
          },
          colorScheme: {
            read: (value) => value ?? "light",
          },
        },
      },
      ConsumerAccount: {
        fields: {
          usdWallet: {
            read: (_, { readField, cache }): Wallet | undefined => {
              const wallets = getWallets({ readField, cache })
              if (wallets === undefined || wallets.length === 0) {
                return undefined
              }

              // TODO: return toReference instead
              // https://www.apollographql.com/docs/react/caching/advanced-topics#cache-redirects
              return wallets.find(
                (wallet) => wallet.walletCurrency === WalletCurrency.Usd,
              )
            },
          },
          btcWallet: {
            read: (_, { readField, cache }) => {
              const wallets = getWallets({ readField, cache })
              if (wallets === undefined || wallets.length === 0) {
                return undefined
              }

              // TODO: return toReference instead
              // https://www.apollographql.com/docs/react/caching/advanced-topics#cache-redirects
              return wallets.find(
                (wallet) => wallet.walletCurrency === WalletCurrency.Btc,
              )
            },
          },
          defaultWallet: {
            read: (_, { readField, cache }): Wallet | undefined => {
              const wallets = getWallets({ readField, cache })
              if (wallets === undefined || wallets.length === 0) {
                return undefined
              }

              const defaultWalletId = readField("defaultWalletId")

              // TODO: return toReference instead
              // https://www.apollographql.com/docs/react/caching/advanced-topics#cache-redirects
              return wallets.find((wallet) => wallet.id === defaultWalletId)
            },
          },
          transactions: relayStylePagination(),
        },
      },
      Wallet: {
        fields: {
          transactions: relayStylePagination(),
        },
      },
    },
  })
