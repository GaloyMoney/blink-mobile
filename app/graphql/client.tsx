import { GraphQLWsLink } from "@apollo/client/link/subscriptions"
import { createClient } from "graphql-ws"

import {
  ApolloClient,
  ApolloLink,
  ApolloProvider,
  HttpLink,
  NormalizedCacheObject,
  ServerError,
  split,
} from "@apollo/client"
import AsyncStorage from "@react-native-async-storage/async-storage"
import DeviceInfo from "react-native-device-info"

import { setContext } from "@apollo/client/link/context"
import { RetryLink } from "@apollo/client/link/retry"
import { getMainDefinition } from "@apollo/client/utilities"

import { createPersistedQueryLink } from "@apollo/client/link/persisted-queries"
import { BUILD_VERSION } from "@app/config"
import { useAppConfig } from "@app/hooks"
import { AsyncStorageWrapper, CachePersistor } from "apollo3-cache-persist"
import jsSha256 from "js-sha256"
import React, { PropsWithChildren, useEffect, useState } from "react"
import { createCache } from "./cache"

import { useI18nContext } from "@app/i18n/i18n-react"
import { isIos } from "../utils/helper"
import { loadString, saveString } from "../utils/storage"
import { AnalyticsContainer } from "./analytics"
import { useLanguageQuery, useRealtimePriceQuery } from "./generated"
import { IsAuthedContextProvider, useIsAuthed } from "./is-authed-context"

import { onError } from "@apollo/client/link/error"

import { getLanguageFromString, getLocaleFromLanguage } from "@app/utils/locale-detector"
import { NetworkErrorToast } from "./network-error-toast"
import { MessagingContainer } from "./messaging"

const noRetryOperations = [
  "intraLedgerPaymentSend",
  "intraLedgerUsdPaymentSend",

  "lnInvoiceFeeProbe",
  "lnInvoicePaymentSend",
  "lnNoAmountInvoiceFeeProbe",
  "lnNoAmountInvoicePaymentSend",
  "lnNoAmountUsdInvoiceFeeProbe",
  "lnUsdInvoiceFeeProbe",
  "lnNoAmountUsdInvoicePaymentSend",

  "onChainPaymentSend",
  "onChainTxFee",

  // no need to retry to upload the token
  // specially as it's running on app start
  // and can create some unwanted loop when token is not valid
  "deviceNotificationTokenCreate",
]

const getAuthorizationHeader = (token: string): string => {
  return `Bearer ${token}`
}

const GaloyClient: React.FC<PropsWithChildren> = ({ children }) => {
  const { appConfig, saveToken } = useAppConfig()
  const [networkError, setNetworkError] = useState<ServerError>()

  const [apolloClient, setApolloClient] = useState<{
    client: ApolloClient<NormalizedCacheObject>
    isAuthed: boolean
  }>()

  useEffect(() => {
    ;(async () => {
      const token = appConfig.token

      console.log(
        `creating new apollo client, token: ${Boolean(token)}, uri: ${
          appConfig.galoyInstance.graphqlUri
        }`,
      )

      const httpLink = new HttpLink({
        uri: appConfig.galoyInstance.graphqlUri,
      })

      // persistedQuery provide client side bandwidth optimization by returning a hash
      // of the uery instead of the whole query
      //
      // use the following line if you want to deactivate in dev
      // const persistedQueryLink = httpLink
      //
      // we are using "js-sha256" because crypto-hash has compatibility issue with react-native
      // from "@apollo/client/link/persisted-queries/types" but not exporterd
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      type SHA256Function = (...args: any[]) => string | PromiseLike<string>
      const sha256: SHA256Function = jsSha256 as unknown as SHA256Function
      const persistedQueryLink = createPersistedQueryLink({ sha256 })

      const authLink = setContext((request, { headers }) => ({
        headers: {
          ...headers,
          authorization: token ? getAuthorizationHeader(token) : "",
        },
      }))

      const retryLink = new RetryLink({
        attempts: {
          max: 5,
          retryIf: (error, operation) => {
            console.debug(JSON.stringify(error), "retry on error")
            return Boolean(error) && !noRetryOperations.includes(operation.operationName)
          },
        },
      })

      const wsLink = new GraphQLWsLink(
        createClient({
          url: appConfig.galoyInstance.graphqlWsUri,
          connectionParams: token
            ? { Authorization: getAuthorizationHeader(token) }
            : undefined,
          retryAttempts: 12,
          shouldRetry: (errOrCloseEvent) => {
            console.warn(
              { errOrCloseEvent },
              "entering shouldRetry function for websocket",
            )
            // TODO: understand how the backend is closing the connection
            // for instance during a new version rollout or k8s upgrade
            //
            // in the meantime:
            // returning true instead of the default 'Any non-`CloseEvent`'
            // to force createClient to attempt a reconnection
            return true
          },
          // Voluntary not using: webSocketImpl: WebSocket
          // seems react native already have an implement of the websocket?
          //
          // TODO: implement keepAlive and reconnection?
          // https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#keepalive
        }),
      )

      const errorLink = onError(({ graphQLErrors, networkError }) => {
        // graphqlErrors should be managed locally
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) => {
            if (message === "PersistedQueryNotFound") {
              console.log(`[GraphQL info]: Message: ${message}, Path: ${path}}`, {
                locations,
              })
            } else {
              console.warn(`[GraphQL error]: Message: ${message}, Path: ${path}}`, {
                locations,
              })
            }
          })
        // only network error are managed globally
        if (networkError) {
          console.log(`[Network error]: ${networkError}`)
          setNetworkError(networkError as ServerError)
        }
      })

      const link = split(
        ({ query }) => {
          const definition = getMainDefinition(query)
          return (
            definition.kind === "OperationDefinition" &&
            definition.operation === "subscription"
          )
        },
        wsLink,
        ApolloLink.from([retryLink, authLink, persistedQueryLink, errorLink, httpLink]),
      )

      const cache = createCache()

      const persistor = new CachePersistor({
        cache,
        storage: new AsyncStorageWrapper(AsyncStorage),
        debug: __DEV__,

        persistenceMapper: async (_data) => {
          // TODO:
          // we should only store the last 20 transactions to keep the cache small
          // there could be other data to filter as well
          // filter your cached data and queries
          // return filteredData
          return _data
        },
      })

      const readableVersion = DeviceInfo.getReadableVersion()

      const client = new ApolloClient({
        cache,
        link,
        name: isIos ? "iOS" : "Android",
        version: readableVersion,
        connectToDevTools: true,
      })

      const buildVersion = DeviceInfo.getBuildNumber()

      // Read the current version from AsyncStorage.
      const currentVersion = await loadString(BUILD_VERSION)

      if (currentVersion === buildVersion) {
        // If the current version matches the latest version,
        // we're good to go and can restore the cache.
        await persistor.restore()
      } else {
        // Otherwise, we'll want to purge the outdated persisted cache
        // and mark ourselves as having updated to the latest version.

        // init the DB. will be override if a cache exists
        await persistor.purge()
        await saveString(BUILD_VERSION, buildVersion)
      }

      client.onClearStore(persistor.purge)

      setApolloClient({ client, isAuthed: Boolean(token) })

      return () => client.cache.reset()
    })()
  }, [appConfig.token, appConfig.galoyInstance, saveToken])

  // Before we show the app, we have to wait for our state to be ready.
  // In the meantime, don't render anything. This will be the background
  // color set in native by rootView's background color.
  //
  // This step should be completely covered over by the splash screen though.
  //
  // You're welcome to swap in your own component to render if your boot up
  // sequence is too slow though.
  if (!apolloClient) {
    return <></>
  }

  return (
    <ApolloProvider client={apolloClient.client}>
      <IsAuthedContextProvider value={apolloClient.isAuthed}>
        <MessagingContainer />
        <NetworkErrorToast networkError={networkError} />
        <LanguageSync />
        <AnalyticsContainer />
        <MyPriceUpdates />
        {children}
      </IsAuthedContextProvider>
    </ApolloProvider>
  )
}

const MyPriceUpdates = () => {
  const isAuthed = useIsAuthed()

  const pollInterval = 5 * 60 * 1000 // 5 min
  useRealtimePriceQuery({
    // only fetch after pollInterval
    // the first query is done by the home page automatically
    fetchPolicy: "cache-only",
    nextFetchPolicy: "network-only",
    pollInterval,
    skip: !isAuthed,
  })

  return null
}

const LanguageSync = () => {
  const isAuthed = useIsAuthed()

  const { data } = useLanguageQuery({ fetchPolicy: "cache-first", skip: !isAuthed })

  const userPreferredLocale = getLocaleFromLanguage(
    getLanguageFromString(data?.me?.language),
  )
  const { locale, setLocale } = useI18nContext()

  useEffect(() => {
    if (userPreferredLocale !== locale) {
      setLocale(userPreferredLocale)
    }
    // setLocale is not set as a dependency because it changes every render
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userPreferredLocale, locale])

  return <></>
}

export { GaloyClient }
