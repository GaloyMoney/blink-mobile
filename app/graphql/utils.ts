import { GraphQLError } from "graphql"
import { GraphQlApplicationError } from "./generated"

export const joinErrorsMessages = (
  errors: readonly GraphQLError[] | readonly GraphQlApplicationError[],
) => {
  return errors.map((err) => err.message).join(", ")
}
