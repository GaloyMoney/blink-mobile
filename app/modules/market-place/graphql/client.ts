import {
  ApolloClient,
  ApolloLink,
  DefaultOptions,
  from,
  HttpLink,
  InMemoryCache,
  fromPromise,
} from "@apollo/client"

import { getMainDefinition } from "@apollo/client/utilities"
import { setContext } from "@apollo/client/link/context"
import { onError } from "@apollo/client/link/error"
import { createUploadLink } from "apollo-upload-client"
import { scriptHostname } from "../utils/helper"

const attemps = 0
const GRAPHQL_REGTEST_URI = `http://${scriptHostname()}:4002/graphql`
const GRAPHQL_TESTNET_URI = "https://api.pvbtc1.staging.pvbtc.cloud/graphql"
const GRAPHQL_MAINNET_URI = "https://api.production.pvbtc.cloud/graphql"

const GRAPHQL_REGTEST_WS_URI = `ws://${scriptHostname()}:4002/graphql`
const GRAPHQL_TESTNET_WS_URI = "wss://api.pvbtc1.staging.pvbtc.cloud/graphql"
const GRAPHQL_MAINNET_WS_URI = "wss://api.production.pvbtc.cloud/graphql"
const errorLink = onError(({ graphQLErrors, networkError, operation }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(async ({ message, locations, path }) => {
      switch (message) {
        case "Unauthorized": {
        }
        default:
          console.log(``)
      }
    })
  if (networkError) console.log(`[Network error]: ${networkError}`)
})
const httpLink = createUploadLink({ uri: "https://api-puravida.herokuapp.com/graphql" })
// new HttpLink({
//   uri: `https://api-puravida.herokuapp.com/graphql`,
// })

const authLink = setContext(async (_, { headers }) => {
  // get the authentication token from local storage if it exists
  // const token = localStorage.getItem('token');
  const token = ""
  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      token: token ? `${token}` : "",
    },
  }
})

export const cache = new InMemoryCache({ addTypename: false })

const defaultOptions: DefaultOptions = {
  watchQuery: {
    fetchPolicy: "no-cache",
    errorPolicy: "ignore",
  },
  query: {
    fetchPolicy: "no-cache",
    errorPolicy: "all",
  },
}

const client = new ApolloClient({
  // link: authLink.concat(errorLink.concat(Â©httpLink)),
  link: from([authLink, errorLink, httpLink]),
  cache,
  defaultOptions,
})

export default client
