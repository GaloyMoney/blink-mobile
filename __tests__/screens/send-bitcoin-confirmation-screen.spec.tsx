import * as React from "react"
import { MockedProvider } from "@apollo/client/testing"
import { InMemoryCache } from "@apollo/client"
import { act, cleanup, fireEvent, render } from "@testing-library/react-native"
import renderer from "react-test-renderer"
import "@testing-library/jest-native/extend-expect"
import "@node_modules/react-native-gesture-handler/jestSetup.js"

import { translate } from "@app/i18n/translate"
import "@mocks/react-native-firebase"
import "@mocks/react-navigation-native"
import {
  LIGHTNING_PAY,
  PAY_KEYSEND_USERNAME,
  SendBitcoinConfirmationScreen,
} from "@app/screens/send-bitcoin-screen"
import { waitForNextRender } from "../helpers/wait"
import { cacheNodeStats, cachePrice, cacheWallet } from "../helpers/cache"
import { IPaymentType } from "@app/utils/parsing"

const payKeysendUsernameMocks = [
  {
    request: {
      query: PAY_KEYSEND_USERNAME,
      variables: {
        amount: 68626,
        destination: "",
        username: "Bitcoin",
        memo: null,
      },
    },
    result: {
      data: {
        invoice: {
          payKeysendUsername: "success",
          __typename: "Invoice",
        },
      },
    },
  },
]

const lightningPayMocks = [
  {
    request: {
      query: LIGHTNING_PAY,
      variables: {
        invoice:
          "lnbc6864270n1p05zvjjpp5fpehvlv3dd2r76065r9v0l3n8qv9mfwu9ryhvpj5xsz3p4hy734qdzhxysv89eqyvmzqsnfw3pxcmmrddpx7mmdypp8yatwvd5zqmmwypqh2em4wd6zqvesyq5yyun4de3ksgz0dek8j2gcqzpgxqrrss6lqa5jllvuglw5tpsug4s2tmt5c8fnerr95fuh8htcsyx52cp3wzswj32xj5gewyfn7mg293v6jla9cz8zndhwdhcnnkul2qkf6pjlspj2nl3j",
        memo: null,
      },
    },
    result: {
      data: {
        invoice: {
          payInvoice: "success",
          __typename: "Invoice",
        },
      },
    },
  },
]

const usernameSatAmount: MoneyAmount = {
  value: 68626,
  currency: "BTC",
}

const usernameRouteParams = {
  address: null,
  amountless: false,
  invoice: null,
  memo: null,
  paymentType: "username" as IPaymentType,
  primaryCurrency: "BTC" as CurrencyType,
  referenceAmount: usernameSatAmount,
  sameNode: null,
  username: "Bitcoin",
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const mNavigation: any = {
  goBack: jest.fn(),
  navigate: jest.fn(),
  pop: jest.fn(),
}

const lightningSatAmount: MoneyAmount = {
  value: 686427,
  currency: "BTC",
}

const lightningRouteParams = {
  address: null,
  amountless: false,
  invoice:
    "lnbc6864270n1p05zvjjpp5fpehvlv3dd2r76065r9v0l3n8qv9mfwu9ryhvpj5xsz3p4hy734qdzhxysv89eqyvmzqsnfw3pxcmmrddpx7mmdypp8yatwvd5zqmmwypqh2em4wd6zqvesyq5yyun4de3ksgz0dek8j2gcqzpgxqrrss6lqa5jllvuglw5tpsug4s2tmt5c8fnerr95fuh8htcsyx52cp3wzswj32xj5gewyfn7mg293v6jla9cz8zndhwdhcnnkul2qkf6pjlspj2nl3j",
  memo: null,
  paymentType: "lightning" as IPaymentType,
  primaryCurrency: "BTC" as CurrencyType,
  referenceAmount: lightningSatAmount,
  sameNode: false,
  username: null,
}

afterEach(cleanup)

describe("SendBitcoinConfirmationScreen", () => {
  const cache = new InMemoryCache()
  cacheNodeStats(cache)
  cachePrice(cache)

  it("render matches snapshot", () => {
    cacheWallet(cache, 1175855)
    const tree = renderer
      .create(
        <MockedProvider cache={cache}>
          <SendBitcoinConfirmationScreen
            navigation={mNavigation}
            route={{
              key: "",
              name: "sendBitcoinConfirmation",
              params: usernameRouteParams,
            }}
          />
        </MockedProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("does not send payment when total exceeds the balance", async () => {
    cacheWallet(cache, 5855)
    const { queryByText } = render(
      <MockedProvider mocks={payKeysendUsernameMocks} cache={cache}>
        <SendBitcoinConfirmationScreen
          navigation={mNavigation}
          route={{
            key: "",
            name: "sendBitcoinConfirmation",
            params: usernameRouteParams,
          }}
        />
      </MockedProvider>,
    )

    expect(
      queryByText(translate("SendBitcoinConfirmationScreen.confirmPayment")),
    ).toBeNull()
  })

  it("successfully sends payment by payKeysendUsername", async () => {
    cacheWallet(cache, 1175855)
    const { getByText, queryByText } = render(
      <MockedProvider mocks={payKeysendUsernameMocks} cache={cache}>
        <SendBitcoinConfirmationScreen
          navigation={mNavigation}
          route={{
            key: "",
            name: "sendBitcoinConfirmation",
            params: usernameRouteParams,
          }}
        />
      </MockedProvider>,
    )

    const sendButton = getByText(
      translate("SendBitcoinConfirmationScreen.confirmPayment"),
    )
    fireEvent.press(sendButton)

    await act(await waitForNextRender)
    expect(queryByText(translate("SendBitcoinScreen.success"))).not.toBeNull()
  })

  it("successfully sends lightning payment with `lightning:` prefix", async () => {
    cacheWallet(cache, 1175855)
    const { getByText, queryByText } = render(
      <MockedProvider mocks={lightningPayMocks} cache={cache}>
        <SendBitcoinConfirmationScreen
          navigation={mNavigation}
          route={{
            key: "",
            name: "sendBitcoinConfirmation",
            params: lightningRouteParams,
          }}
        />
      </MockedProvider>,
    )
    const sendButton = getByText(
      translate("SendBitcoinConfirmationScreen.confirmPayment"),
    )
    fireEvent.press(sendButton)

    await act(await waitForNextRender)
    expect(queryByText(translate("SendBitcoinScreen.success"))).not.toBeNull()
  })
})
