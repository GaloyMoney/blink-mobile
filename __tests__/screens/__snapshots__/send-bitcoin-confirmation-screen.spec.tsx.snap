// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendBitcoinConfirmationScreen render matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#E6EBEf",
          "flex": 1,
          "height": "100%",
        },
        Object {},
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection={
      Array [
        "down",
      ]
    }
    swipeThreshold={50}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      deviceHeight={null}
      deviceWidth={null}
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "justifyContent": "center",
          "left": 0,
          "margin": 37.5,
          "marginBottom": 0,
          "marginHorizontal": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection={
        Array [
          "down",
        ]
      }
      swipeThreshold={50}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
      <RNCSafeAreaView
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "paddingHorizontal": 20,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Object {
              "height": 14,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#CFD9E2",
                  "fontSize": 72,
                },
                Object {
                  "height": 40,
                  "top": -40,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ï“‰
          </Text>
        </View>
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginVertical": 8,
            }
          }
        >
          
        </Text>
        <View
          style={
            Object {
              "alignContent": "stretch",
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 3,
                },
                undefined,
                false,
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#3050C4",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginHorizontal": 20,
                    "marginVertical": 10,
                    "padding": 8,
                    "width": 145,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Dismiss
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 3,
                },
                undefined,
                false,
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#3050C4",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginHorizontal": 20,
                    "marginVertical": 10,
                    "padding": 8,
                    "width": 145,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Open
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RNCSafeAreaView>
    </View>
  </Modal>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "height": "100%",
          "justifyContent": "flex-start",
          "width": "100%",
        },
        Object {},
      ]
    }
  >
    <View>
      <View>
        <View>
          <View>
            <Text>
              To:
            </Text>
            <Text>
              Bitcoin
            </Text>
          </View>
          <View>
            <Text>
              Amount:
            </Text>
            <Text>
              68,626 sats
            </Text>
            <Text>
              $27.22
            </Text>
          </View>
          <View>
            <Text>
              Fee:
            </Text>
            <Text>
              0 sats
            </Text>
          </View>
        </View>
      </View>
      <View />
      <View>
        <View>
          <Text>
            Do you want to confirm this payment?
          </Text>
          <Text>
            Payments are final.
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "borderRadius": 3,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 3,
              },
              undefined,
              false,
            ]
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "busy": false,
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#2089dc",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontSize": 18,
                    "paddingVertical": 1,
                    "textAlign": "center",
                  }
                }
              >
                Confirm payment
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
